gui: Display warnings as rich text

hebasto opened the pull request:
On master (6621be53517d69ab855cee4a5978a44d6a133ba3), warnings that contain `<hr />` HTML tag are not displayed correctly:

![Screenshot from 2020-05-06 11-30-10](https://user-images.githubusercontent.com/32963518/81177281-0e49fc80-8faf-11ea-8cac-8847aa517e86.png)

Fixed:

![Screenshot from 2020-05-07 07-30-48](https://user-images.githubusercontent.com/32963518/81255618-ca9ad580-9036-11ea-90ad-7f4d89c1880d.png)

vasild commented on src/qt/bitcoin.cpp:
```diff
@@ -135,7 +135,7 @@ BitcoinCore::BitcoinCore(interfaces::Node& node) :
 {
     PrintExceptionContinue(e, "Runaway exception");
-    Q_EMIT runawayException(QString::fromStdString(m_node.getWarnings()));
+    Q_EMIT runawayException(QString::fromStdString(m_node.getWarnings()).replace("<hr />", "\n\n"));
```
So we have

https://github.com/bitcoin/bitcoin/blob/88b2652fadf6e004e751d48884ae8d4cf5c452b8/src/warnings.h#L17

https://github.com/bitcoin/bitcoin/blob/88b2652fadf6e004e751d48884ae8d4cf5c452b8/src/warnings.cpp#L45

Would it be better to extract the `warning_separator` variable into a global constant and use it here instead of hardcoding `"<hr />"` in both places?

Even better would be to return a vector of strings instead of multiple strings concatenated to each other with `"<hr />"`, but that would be a much bigger change, I am not suggesting to do it here in this PR.

hebasto replied:
> ... a much bigger change...

That was my concern too :)

promag replied:
How about drop `warning_separator` and use `std::endl`?

AFAICT this `<hr />` was introduced in #7579.

hebasto replied:
[Reworked](https://github.com/bitcoin/bitcoin/pull/18898#issuecomment-625027678).

vasild reviewed (commented):
Should the following be adjusted too?

https://github.com/bitcoin/bitcoin/blob/88b2652fadf6e004e751d48884ae8d4cf5c452b8/src/qt/clientmodel.cpp#L145

0xB10C commented:
Is there a specific way to trigger a runaway exception?

hebasto commented:
> Is there a specific way to trigger a runaway exception?

By mutating the code :)
E.g., https://github.com/bitcoin/bitcoin/issues/18643#issue-599952770

Empact commented:
Concept ACK

Took a crack at an alternative and it seems reasonable to me:
https://github.com/bitcoin/bitcoin/compare/master...Empact:200506-no-html

hebasto committed:
qt: Display warnings as rich text

hebasto commented:
Updated 1b87b55f08ed11e5aba63e49397cf294aaf3ff8f -> a9d28afe23a94efdccc53f9f10716f3a0c9337eb ([pr18898.01](https://github.com/hebasto/bitcoin/commits/pr18898.01) -> [pr18898.02](https://github.com/hebasto/bitcoin/commits/pr18898.02), [diff](https://github.com/hebasto/bitcoin/compare/pr18898.01..pr18898.02)):

- reworked to the minimal diff

---

The OP has been updated.

hebasto commented:
Reopened to update Travis CI build list.

jonasschnelli commented:
Does simply replacing the newlines at the end with a HTML `<br />` fix the `<tr/>` problem?

hebasto commented:
@jonasschnelli 
> Does simply replacing the newlines at the end with a HTML `<br />` fix the `<tr/>` problem?

Sorry. Mind rewording your question?

jonasschnelli commented:
> Sorry. Mind rewording your question?

Sorry if it was not clear.
You're only changing the `\n` to `<br />`. You write in the PR description that is solves the `<hr />` issue (display a line instead of the plaintext html tag). I wonder why adding a `<br>` should make it rich text. Can you explain?

hebasto commented:
> > Sorry. Mind rewording your question?
> 
> Sorry if it was not clear.
> You're only changing the `\n` to `<br />`. You write in the PR description that is solves the `<hr />` issue (display a line instead of the plaintext html tag). I wonder why adding a `<br>` should make it rich text. Can you explain?

From `QMessageBox` class [docs](https://doc.qt.io/qt-5/qmessagebox.html#rich-text-and-the-text-format-property):
> The main text and informative text properties can be either plain text or rich text. These strings are interpreted according to the setting of the text format property. The default setting is auto-text.

As we do not have means to set `Qt::RichText` instead of `Qt::AutoText` directly, replacing s/`\n`/`<br>`/ forces the text string to be interpreted as a rich text.

promag commented:
Also https://doc.qt.io/qt-5/qt-sub-qtgui.html#mightBeRichText.

jonasschnelli commented:
Now I understand.
utACK a9d28afe23a94efdccc53f9f10716f3a0c9337eb

promag reviewed (commented):
Code review ACK a9d28afe23a94efdccc53f9f10716f3a0c9337eb.
